src/librustc_infer/infer/combine.rs

```diff
@@ -649,13 +649,17 @@ impl TypeRelation<'tcx> for Generalizer<'_, 'tcx> {
     ) -> RelateResult<'tcx, &'tcx ty::Const<'tcx>> {
         assert_eq!(c, c2); // we are abusing TypeRelation here; both LHS and RHS ought to be ==
 
+        debug!("generalize: consts c={:?}", c);
         match c.val {
             ty::ConstKind::Infer(InferConst::Var(vid)) => {
                 let mut inner = self.infcx.inner.borrow_mut();
                 let variable_table = &mut inner.const_unification_table();
                 let var_value = variable_table.probe_value(vid);
                 match var_value.val {
-                    ConstVariableValue::Known { value: u } => self.relate(&u, &u),
+                    ConstVariableValue::Known { value: u } => {
+                        drop(inner);
```

Comment:
Why does `inner` exist at all here? It doesn't in `fn types`. Seems to me like nothing should actually keep references into `const_unification_table`. It's safer to just have very-short-scoped `borrow()`/`borrow_mut()`s.

Reply:
Probably so we don't have to reborrow `infcx.inner` in the `ConstVariableValue` branch :thinking: 

Implemented your suggestion